debian/grub/25_linux_acrn

```diff
@@ -97,117 +123,171 @@ fi
+
+        # start with kernel mod tag
+        tagname="$(eval echo \${ACRN_KERN_MOD_VM${id}})"
+        if [ -z "${tagname}" ]; then
+           gettext_printf "ACRN: No kernel module tag set for %s(%s), skipping\n" "$(eval echo \${ACRN_NAME_VM${id}})" "VM${id}" >&2
+           continue
+        fi
+        if [ -z "$(eval echo \${GRUB_ACRN_MOD_${tagname}})" ]; then
+           gettext_printf "ACRN: GRUB_ACRN_MOD_${tagname} not set, skipping %s(%s)\n" "$(eval echo \${ACRN_NAME_VM${id}})" "VM${id}" >&2
+           continue
+        fi
+        gettext_printf "Adding ACRN pre-launched VM %s(%s): %s\n" "$(eval echo \${ACRN_NAME_VM${id}})" "VM${id}" "$(eval echo \${GRUB_ACRN_MOD_${tagname}})">&2
+
+        message="$(gettext_printf "Loading ACRN %s kernel for %s: %s ..." "VM${id}" "$(eval echo \${ACRN_NAME_VM${id}})" "$(eval echo \${GRUB_ACRN_MOD_${tagname}})")"
+        basename=$(basename $(eval echo \${GRUB_ACRN_MOD_${tagname}}))
+        dirname=$(dirname $(eval echo \${GRUB_ACRN_MOD_${tagname}}))
+        rel_dirname=$(make_system_path_relative_to_its_root $dirname)
+        cat << EOF
+	echo	'$(echo "$message" | grub_quote)'
+	${module_loader}	${rel_dirname}/${basename} ${tagname} $(eval echo \${GRUB_ACRN_MOD_CMDLINE_${tagname}})
+EOF
+        # next: use ACPI mod tag
+        tagname="$(eval echo \${ACRN_ACPI_MOD_VM${id}})"
+        if [ -n "$(eval echo \${GRUB_ACRN_MOD_${tagname}})" ]; then
+            message="$(gettext_printf "Loading ACRN %s ACPI data for %s: %s ..." "VM${id}" "$(eval echo \${ACRN_NAME_VM${id}})" "$(eval echo \${GRUB_ACRN_MOD_${tagname}})")"
+            basename=$(basename $(eval echo \${GRUB_ACRN_MOD_${tagname}}))
+            dirname=$(dirname $(eval echo \${GRUB_ACRN_MOD_${tagname}}))
+            rel_dirname=$(make_system_path_relative_to_its_root $dirname)
+            cat << EOF
+	echo	'$(echo "$message" | grub_quote)'
+	${module_loader}	${rel_dirname}/${basename} ${tagname}
+EOF
+        else
+            gettext_printf "ACRN: GRUB_ACRN_MOD_${tagname} not set, skipping\n" >&2
+        fi
+        # finally care for ramdisk module tag
+        tagname="$(eval echo \${ACRN_RAMDISK_MOD_VM${id}})"
+        if [ -n "${tagname}" ]; then
+            if [ -n "$(eval echo \${GRUB_ACRN_MOD_${tagname}})" ]; then
+                message="$(gettext_printf "Loading ACRN %s ramdisk for %s: %s ..." "VM${id}" "$(eval echo \${ACRN_NAME_VM${id}})" "$(eval echo \${GRUB_ACRN_MOD_${tagname}})")"
+                basename=$(basename $(eval echo \${GRUB_ACRN_MOD_${tagname}}))
+                dirname=$(dirname $(eval echo \${GRUB_ACRN_MOD_${tagname}}))
+                rel_dirname=$(make_system_path_relative_to_its_root $dirname)
+                cat << EOF
+	echo	'$(echo "$message" | grub_quote)'
+	${module_loader} --nounzip	${rel_dirname}/${basename} ${tagname}
+EOF
+            else
+                gettext_printf "ACRN: GRUB_ACRN_MOD_${tagname} not set, skipping\n" >&2
```

Comment:
The code snippet that generates the module loader seems to repeat 3 times with differences only in the tagname and format string being used. Is it possible to define a function to avoid code repetition?

Reply:
You're perfectly right, this code repetition should be avoided! I'll care for that!